### PR TITLE
Improve mypy coverage by adding --namespace-packages

### PR DESCRIPTION
**Pull request by chriselion**
Turning on --namespace-packages gets mypy to recognize more types across files. Unfortunately it also triggers an assert in mypy, because it ends up reporting the same file twice (see https://github.com/python/mypy/issues/7510).

I haven't seen any progress on the mypy issue, so I forked, disabled the assertion, and set up precommit to look at the fork instead.

This PR fixes the errors uncovered by doing this.

### CODE REVIEW

File: ml-agents/mlagents/trainers/trainer.py [hunk 1]
```diff
@@ -282,3 +282,8 @@ def update_policy(self):
         Uses demonstration_buffer to update model.
         """
         raise UnityTrainerException("The update_model method was not implemented.")
+
+    def clear_update_buffer(self) -> None:
+        raise UnityTrainerException(
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/trainer.py]: @ervteng Not sure what we should do about this. This line (correctly) complains
https://github.com/Unity-Technologies/ml-agents/blob/40d74dc1125f69b0af7e14f216a6dcea87e20b5c/ml-agents/mlagents/trainers/trainer_controller.py#L298
because clear_update_buffer() was only on RLTrainer, not Trainer. I figure it might be a moot point after the BC removal - is part of that PR combining Trainer and RLTrainer?

File: ml-agents-envs/mlagents/envs/environment.py [hunk 2]
```diff
@@ -494,7 +495,7 @@ def _get_state(self, output: UnityRLOutputProto) -> AllBrainInfo:
 
     @staticmethod
     def _parse_side_channel_message(
-        side_channels: Dict[int, SideChannel], data: bytearray
+        side_channels: Dict[int, SideChannel], data: bytes
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/environment.py]: `bytes` is immutable - these seemed better for processing inputs. I left outputs as` bytearray`s though.

File: ml-agents-envs/mlagents/envs/brain.py [hunk 3]
```diff
@@ -160,7 +160,9 @@ def process_pixels(image_bytes: bytes, gray_scale: bool) -> np.ndarray:
     @timed
     def from_agent_proto(
         worker_id: int,
-        agent_info_list: List[AgentInfoProto],
+        agent_info_list: Collection[
+            AgentInfoProto
+        ],  # pylint: disable=unsubscriptable-object
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/brain.py]: See https://github.com/PyCQA/pylint/issues/3292 :sigh:

File: ml-agents/mlagents/trainers/trainer.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer.py]: The BC remove doesn't combine the two - I think the right thing is to put it only in rl_trainer, since there's no "update buffer" per se for all trainers

Comment (chriselion) [reply ml-agents/mlagents/trainers/trainer.py]: The what should the TrainerController call? It only knows about Trainers, not RLTrainers.

File: ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py [hunk 4]
```diff
@@ -52,7 +52,14 @@ def list_properties(self) -> List[str]:
         Returns a list of all the string identifiers of the properties
         currently present in the Unity Environment.
         """
-        return self._float_properties.keys()
+        return list(self._float_properties.keys())
+
+    def get_property_dict(self) -> Dict[str, float]:
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py]: Add this since the usage pattern was always list_properties then get_property on each.

File: ml-agents/mlagents/trainers/agent_processor.py [hunk 5]
```diff
@@ -28,7 +28,7 @@ def reset_local_buffers(self) -> None:
     def append_to_update_buffer(
         self,
         update_buffer: AgentBuffer,
-        agent_id: str,
+        agent_id: int,
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/agent_processor.py]: @ervteng Is this correct now? mypy was complaining about demo_loader calling this with `0` instead of a string.

File: ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py (hunk 4 above)
Comment (vincentpierre) [reply ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py]: Do we still need get and list properties then?
Also, it is not clear that modifying this dict will not modify the unity environment...

File: ml-agents/mlagents/trainers/models.py [hunk 6]
```diff
@@ -88,6 +88,10 @@ def __init__(
         self.running_variance: Optional[tf.Variable] = None
         self.update_normalization: Optional[tf.Operation] = None
         self.value: Optional[tf.Tensor] = None
+        self.all_log_probs: Optional[tf.Tensor] = None
+        self.output: Optional[tf.Tensor] = None
+        self.selected_actions: Optional[tf.Tensor] = None
+        self.action_holder: Optional[tf.Tensor] = None
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/models.py]: Did you intend to add those? It does not seem related to mypy coverage.

Comment (chriselion) [reply ml-agents/mlagents/trainers/models.py]: Yes, there were places where mypy was complaining (correctly) that e.g. all_log_probs wasn't a member of LearningModel. In practice it was OK because they were always defined in PPOModel or SACModel. This just defines them in the base class too.

File: ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py (hunk 4 above)
Comment (chriselion) [reply ml-agents-envs/mlagents/envs/side_channel/float_properties_channel.py]: Not sure we need list. I think get() is still useful. 

Would `get_property_dict_copy()` make it clearer? Unfortunately there's no clean way to return a read-only dictionary.

File: ml-agents-envs/mlagents/envs/base_env.py [hunk 7]
```diff
@@ -18,12 +18,12 @@
...
 
 from abc import ABC, abstractmethod
-from typing import List, NamedTuple, Tuple, Optional, Union, Dict, NewType
+from typing import List, NamedTuple, Tuple, Optional, Union, Dict
 import numpy as np
 from enum import Enum
 
-AgentId = NewType("AgentId", int)
-AgentGroup = NewType("AgentGroup", str)
+AgentId = int
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/base_env.py]: I think this was going to be more trouble that it was worth. This still aliases the types so that it's a little clearer than the raw int/str types.

File: ml-agents-envs/mlagents/envs/environment.py [hunk 8]
```diff
@@ -371,8 +372,8 @@ def set_action_for_agent(
             action = action.astype(expected_type)
 
         if agent_group not in self._env_actions:
-            self._env_actions[agent_group] = self._empty_action(
-                spec, self._env_state[agent_group].n_agents()
+            self._env_actions[agent_group] = spec.create_empty_action(
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/environment.py]: This was a bug.

File: ml-agents/mlagents/trainers/agent_processor.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/agent_processor.py]: Demo loader is wrong, it uses an int for the update buffer. However in env_manager that int is turned into a string, so the `int` won't work in the general case.

Comment (ervteng) [reply ml-agents/mlagents/trainers/agent_processor.py]: Actually it's both, in different parts of the code :P

Comment (chriselion) [reply ml-agents/mlagents/trainers/agent_processor.py]: Made it a Union[int, str] for now
